pallets/parachain-staking/src/migrations.rs

```diff
@@ -39,100 +39,98 @@ use alloc::format;
 use frame_support::{
 	migration::storage_key_iter,
 	pallet_prelude::PhantomData,
+	storage,
 	traits::{Get, OnRuntimeUpgrade, ReservableCurrency},
 	weights::Weight,
 };
 #[cfg(feature = "try-runtime")]
 use scale_info::prelude::string::String;
 use sp_runtime::traits::{Saturating, Zero};
 use sp_runtime::Percent;
-use sp_std::collections::btree_set::BTreeSet;
 use sp_std::{convert::TryInto, vec, vec::Vec};
 
-/// Migration `AtStake` storage item to include auto-compound value
+/// Migrate `AtStake` storage item to include auto-compound value for unpaid rounds.
 pub struct MigrateAtStakeAutoCompound<T>(PhantomData<T>);
 impl<T: Config> MigrateAtStakeAutoCompound<T> {
-	const PALLET_PREFIX: &'static [u8] = b"ParachainStaking";
-	const AT_STAKE_PREFIX: &'static [u8] = b"AtStake";
+	/// Get keys for the `AtStake` storage for the rounds up to `RewardPaymentDelay` rounds ago.
+	/// We migrate only the last unpaid rounds due to the presence of stale entries in `AtStake`
+	/// which significantly increase the PoV size.
+	fn unpaid_rounds_keys() -> Vec<(RoundIndex, T::AccountId, Vec<u8>)> {
+		let current_round = <Round<T>>::get().current;
+		let max_unpaid_round = current_round.saturating_sub(T::RewardPaymentDelay::get());
+		(max_unpaid_round..=current_round)
+			.into_iter()
+			.flat_map(|round| {
+				<AtStake<T>>::iter_key_prefix(round).map(move |candidate| {
+					let key = <AtStake<T>>::hashed_key_for(round.clone(), candidate.clone());
+					(round, candidate, key)
+				})
+			})
+			.collect::<Vec<_>>()
```

Comment:
```suggestion

```

Here it is useless to create a Vec and copy the data into it. It is more efficient to return the iterator directly :)